<table align="left" width=100%>
    <tr>
        <td width="20%">
            <img src="GL-2.png">
        </td>
        <td>
            <div align="center">
                <font color="#21618C" size=8px>
                  <b> Final Exam<br> </b>
                </font>
            </div>
        </td>
    </tr>
</table>

# INTRODUCTION TO PYTHON


## Total Marks : 70 Marks

----

# SECTION A: 20 MARKS

 _________________

### Q1.  Write a program that asks the user to enter size of the list, then do the following: (5Marks)
(a) if size of the list is less than 4 then print a message to user "Invalid size and enter a size more than 4". The program will continue from begining.

(b) After taking valid size , program should take list items from user and create user defined list. 

(c) Use the choice() function to print out random item from the list.

(d) Use the sample() function to print out three random items from the list.

(e) Use the shuffle() function to reorder the list and then print out the list.


**Choice(),shuffle(),sample() functions are defined under random library. You need to import random library to use it.**

***
#### Sample Input/Output:


<span style="font-family:consolas ">

Enter number of elements in the list: 3

Invalid size and enter a size more than 4

Enter number of elements in the list: 5

1

2

3

4

5

Random item from the list:  1

Three random items from the list: ['1', '4', '2']

Shuffeled list: ['3', '4', '5', '1', '2']

***

In [19]:
import random
list = []
size = int(input('Enter number of elements in the list: '))
if size<4:
    print('Invalid Size. Size should be atleast 4.')
    while True:
        size = int(input('Rectify and enter number of elements in the list: '))
        if size<4:
            print('Invalid Size. Size should be atleast 4.')
        else:
            break


for i in range(size):
    a = int(input())
    print(a)
    list.append(a)

print(f"Random item from the list: {random.choice(list)}")
print(f"Three random items from the list: {random.sample(list,3)}")
list1 = list
random.shuffle(list1)
print(f"Shuffeled list: {list1}")

Invalid Size. Size should be atleast 4.
1
2
3
4
5
Random item from the list: 5
Three random items from the list: [2, 1, 4]
Shuffeled list: [4, 5, 1, 3, 2]


### Q2. Ask the user for a number and then print the following, where the pattern ends at the number that the user enters.
<pre>
1
 2
  3
   4
    5
   4
  3
 2
1
</pre>

***
#### Sample Input/Output:

***

In [ ]:
n = int(input('Enter the number: '))
s = ' '
for i in range(n):
    print(f"{s*i}{i+1}")
    s = ' '
s = ' '

for i in range(n-1,0,-1):
    print(f"{s*(i-1)}{i}")
    s = ' '


1
 2
  3
   4
    5
   4
  3
 2
1


### Q3.Write a program to validate the data entered by its user and then perform conversion from kilogram to pounds. Write a program that asks the user for a weight and converts it from kilograms to pounds. Whenever the user enters a weight below 0, the program should tell them that their entry is invalid and then ask them again to enter a weight. At the end it should show a weight converted into pounds. do this repeatedly untill user enters 'STOP' [use of IF statement is restricated here] (5 marks)


***
#### Sample Input/Output:

Enter Weight or enter STOP: 10

Weight in pounds:  22.046226218 lbs

Enter Weight: 50

Weight in pounds:  110.23113109 lbs

Enter Weight: 45

Weight in pounds:  99.208017981 lbs

Enter Weight: 62

Weight in pounds:  136.6866025516 lbs

Enter Weight: STOP

***

In [ ]:
while True:
    try:
        weight = float(input('Enter Weight or STOP: '))
        if weight < 0:
            print('Invalid Weight.')
            while True:
                weight = float(input('Rectify and enter the weight: '))
                if weight < 0:
                    print('Invalid Weight.')
                else:
                    break
                
        print(f"Weight in pounds: {round(2.2046226218*weight,10)} lbs")
    except ValueError:
        break

Weight in pounds: 136.6866025516 lbs
Weight in pounds: 110.23113109 lbs
Weight in pounds: 99.208017981 lbs
Invalid Weight.
Invalid Weight.
Weight in pounds: 191.8021680966 lbs


### Q4. Ask the user to enter a list of 10 elemnts containing numbers between 1 and 15. Then replace all of the entries in the list that are greater than 10, with 10 using map() function.(5 Marks)


***
#### Sample Input/Output:

Enter a number between 1 to 15: 3

Enter a number between 1 to 15: 13

Enter a number between 1 to 15: 7

Enter a number between 1 to 15: 12

Enter a number between 1 to 15: 13

Enter a number between 1 to 15: 4

Enter a number between 1 to 15: 5

Enter a number between 1 to 15: 6

Enter a number between 1 to 15: 17

Enter a number between 1 to 15: 18

Output list :  [3, 10, 7, 10, 10, 4, 5, 6, 10, 10]

***

In [ ]:
list1 = []
for i in range(10):
    list1.append(int(input('Enter a number between 1 to 15: ')))

output_list = list(map(lambda x : 10 if x>10 else x, list1))
print('Output List:', output_list)

Output List: [2, 4, 10, 10, 10, 3, 2, 6, 10, 3]


# SECTION B: 20 MARKS

---------------------

### Q5. A superstore has multiple items to sell. These items are categorised as  category A and category B. For festival time the superstore decided to declare some discount, as given below. 

### for category A items, it charges Rs.15 per item if you buy less than 30 quantity of an item. If you buy between 30 and 99 quantity of an item, the cost is Rs.12 per item. If buying quantity is 100 or more for an item, the cost is Rs.10 per item. 

### For category B items, it charges Rs.30 per item if you buy less than 50 quantity of an item. If you buy between 50 and 99 quantity of an item, the cost is Rs.25 per item. If buying quantity is 100 or more for an item, the cost is Rs.20 per item. Further it also gives discount of 5% if total bill exceeds 1000 and 10% if bill exceeds 1500.

### Write a program which repeatedly ask user to enter no of items and its category. If user enters 'STOP' it shows user has done with the purchase, now calculate total payable amount and display it to user(10 Marks)


***
#### Sample Input/Output:
Enter no of items and its category: 23,A

Enter no of items and its category: 56, b

Enter no of items and its category: 67,b

Enter no of items and its category: 110,a

Enter no of items and its category: STOP

Total amount payable is :  2808.0

***

In [ ]:
list = []
while True:
    try:
        items,cat = input('Enter no. of items and its category: ').split(',')
        list.append([cat,int(items)])    
    except ValueError:
        break

count = 0
for items in list:
    if items[0] == 'a' or items[0] == 'A':
        if items[1] <= 30:
            count += items[1]*15
        elif items[1]>30 and items[1]<=99:
            count += items[1]*12
        else:
            count += items[1]*10

    else:
        if items[1] <= 50:
            count += items[1]*30
        elif items[1]>50 and items[1]<=99:
            count += items[1]*25
        else:
            count += items[1]*20

if count >= 1000 and count < 1500:
    count = count*0.95
elif count >= 1500:
    count = count*0.9

print(count)


4068.0


### Q6.Create a 5*5  list of numbers from 1 to 10 using list comprehension. Print the nested list. 
### Create a dictionary whose keys are the numbers and whose values are the frequencies of the number in the nested list mentined above
### Print the dictionary in the descending order of frequencies of the numbers. 
### Then print the three most common numbers.(10 Marks) 


***
#### Sample Input/Output:

5 * 5 nested list : 
[[10, 3, 10, 9, 4], [1, 6, 1, 2, 5], [4, 1, 5, 9, 7], [9, 7, 5, 4, 1], [8, 7, 3, 3, 3]]


Dictionary for number and its frequency : 
{1: 4, 2: 1, 3: 4, 4: 3, 5: 3, 6: 1, 7: 3, 8: 1, 9: 3, 10: 2}


Dictionary in descending order of frequency : 
{1: 4, 3: 4, 4: 3, 5: 3, 7: 3, 9: 3, 10: 2, 2: 1, 6: 1, 8: 1}


Top three frequent numbers are [1, 3, 4]

***

In [ ]:
import random
list = [[i for i in random.sample(range(1,11),5)] for j in range(0,4)]
list1 = []
for i in list:
    for j in i:
        list1.append(j)

dict1 = {}
for i in set(list1):
    dict1[i] = list1.count(i)

print('Dictionary for number and its frequency :',dict1)
dict2 = dict(sorted(dict1.items(), key=lambda x: x[1], reverse=True))
print('Dictionary in Descending order of frequency : ',dict2)
list3 = []
for i in dict2.keys():
    list3.append(i)
print('Top three frequent numbers are',list3[0:3])

Dictionary for number and its frequency : {2: 2, 3: 1, 4: 3, 5: 2, 6: 1, 7: 4, 8: 3, 9: 1, 10: 3}
Dictionary in Descending order of frequency :  {7: 4, 4: 3, 8: 3, 10: 3, 2: 2, 5: 2, 3: 1, 6: 1, 9: 1}
Top three frequent numbers are [7, 4, 8]


In [ ]:
dict = {'a':1,'b':2,'c':3}
for i in dict.items():
    print(i)

('a', 1)
('b', 2)
('c', 3)


# SECTION C: 30 MARKS

--------------------

### Q7. One way of encrypting a string message is to rearrange its characters. Simple way to rearrange the characters is to pick out the characters at even indices, put them first in the encrypted string and follow them by the odd characters.
#### For example, the string message would be encrypted as msaeesg because the even characters are m, s, a, e (at indices 0, 2, 4, and 6) and the odd characters are e, s, g (at indices 1, 3, and 5).
#### (a) Write a function encrypt()to encrypt the string given by user.
#### (b) Write a function decrypt() to decrypt a string that was encrypted before.(15 Marks)


***
#### Sample Input/Output:
<span style="font-family:consolas ">
Encrypted message is :  SceMsaeertesg
    
Decrypted message is :  SecretMessage
***

In [ ]:
def Encrypt(s):
    list1 = []
    for i in s:
        list1.append(i)
    n = len(list1)
    list_even = list1[0:n:2]
    list_odd = list1[1:n:2]
    list_enc = list_even+list_odd
    str = ''
    for i in list_enc:
        str += i
    return str

def Decrypt(s):
    n = len(s)
    str_even = s[0:n//2+(n%2)]
    #print(str_even)
    str_odd = s[n//2+(n%2):n]
    #print(str_odd)
    str = ''
    for i in range(n//2):
        str = str+str_even[i]+str_odd[i]

    if len(str) != n:
        str = str + str_even[-1]
    return str

s = input('Enter the message: ')
print('Encrypted message is :',Encrypt(s))
print('Decrypted message is :',Decrypt(Encrypt(s)))

Encrypted message is : Eccoeinylpda
Decrypted message is : Encyclopedia


### Q8. Write a program that checks to see if a 4 * 4 list is a magic square. In a magic square, every row, column, and the two diagonals add up to the same value. Write a function magic_square which returns True if list is magic square otherwise it returns false (15 Marks)

E.g Given below is 4 * 4 magic Square
<pre>
    1  15  14  4
    10 11   8  5
    7  6    9  12
    16 2    3  13


***
#### Sample Input/Output: 
Its a magic Square

***

In [18]:
num = input('Enter the numbers: ').split(',')
num_list = [int(i) for i in num]

matrix = [num_list[i:i+int(len(num_list)**0.5)] for i in range(0,len(num_list),int(len(num_list)**0.5))]
print(matrix)
count = 0
for i in range(0,len(matrix)):
    if sum(matrix[i]) == sum(matrix[i-1]):
        count += 1

col_1 = [[item[i] for item in matrix] for i in range(len(matrix))]
print(col_1)
for i in range(0,len(matrix)):
    if sum(col_1[i]) == sum(col_1[i-1]):
        count += 1
count
diag_1 = [matrix[i][j] for i in range(len(matrix)) for j in range(len(matrix)) if i==j]
print(diag_1)
diag_2 = [matrix[i][j] for i in range(len(matrix)) for j in range(len(matrix)) if i+j == (len(matrix)-1)]
print(diag_2)

if sum(diag_1) == sum(diag_2):
    count += 1

if count == 2*(len(num_list)**0.5)+1 :
    print(f"It is a {int(len(num_list)**0.5)}x{int(len(num_list)**0.5)} Magic Square.")
else: print('It is not a Magic Square.')

[[18, 25, 2, 9, 11], [4, 6, 13, 20, 22], [15, 17, 24, 1, 8], [21, 3, 10, 12, 19], [7, 14, 16, 23, 5]]
[[18, 4, 15, 21, 7], [25, 6, 17, 3, 14], [2, 13, 24, 10, 16], [9, 20, 1, 12, 23], [11, 22, 8, 19, 5]]
[18, 6, 24, 12, 5]
[11, 20, 24, 3, 7]
It is a 5x5 Magic Square.


In [ ]:
a,b,c = 3,3,3
if a == b == c:
    print('Oh yeah')

Oh yeah


In [1]:
a = 1
b = 2
a+b

3